src/main/scala/li/cil/oc/integration/forestry/UpgradeBeekeeper.scala

```diff
@@ -0,0 +1,197 @@
+    if (host.mainInventory.getSizeInventory > 0) {
+      position.world.get.getTileEntity(position.offset(side)) match {
+        case housing: IBeeHousing => f(housing)
+        case _ => result(false, "Not facing an apiary")
+      }
+    }
+    else result(false, "No inventory?")
+
+  @Callback(doc = """function(side:number):boolean -- Swap the queen from the selected slot with the apiary at the specified side.""")
+  def swapQueen(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => {
+      val selected = host.mainInventory.getStackInSlot(host.selectedSlot)
+      val oldQueen = housing.getBeeInventory.getQueen
+      housing.getBeeInventory.setQueen(selected)
+      host.mainInventory.setInventorySlotContents(host.selectedSlot, oldQueen)
+      result(true)
+    })
+  }
+
+  @Callback(doc = """function(side:number):boolean -- Swap the drone from the selected slot with the apiary at the specified side.""")
+  def swapDrone(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => {
+      val selected = host.mainInventory.getStackInSlot(host.selectedSlot)
+      val oldQueen = housing.getBeeInventory.getDrone
+      housing.getBeeInventory.setDrone(selected)
+      host.mainInventory.setInventorySlotContents(host.selectedSlot, oldQueen)
+      result(true)
+    })
+  }
+
+  @Callback(doc = """function(side:number):number -- Get current progress percent for the apiary at the specified side.""")
+  def getBeeProgress(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => result(housing.getBeekeepingLogic.getBeeProgressPercent))
+  }
+
+  @Callback(doc = """function(side:number):boolean -- Checks if current bee in the apiary at the specified side can work now.""")
+  def canWork(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => result(housing.getBeekeepingLogic.canWork))
+  }
+
+  @Callback(doc = """function(honeyslot:number):boolean -- Analyzes bee in selected slot, uses honey from the specified slot.""")
+  def analyze(context: Context, args: Arguments): Array[AnyRef] = {
+    val inventory = host.mainInventory
+    val specimenSlot = host.selectedSlot
+    val specimen = inventory.getStackInSlot(specimenSlot)
+    if (!BeeManager.beeRoot.isMember(specimen))
+      return result(false, "Not a bee")
+
+    val honeySlot = args.checkSlot(inventory, 0)
+    val honeyStack = inventory.getStackInSlot(honeySlot)
+    if (honeyStack== null || honeyStack.stackSize == 0 || (honeyStack.getItem != PluginApiculture.items.honeydew && honeyStack.getItem != PluginApiculture.items.honeyDrop))
+      return result(false, "No honey!")
+
+    val individual = BeeManager.beeRoot.getMember(specimen)
+    if (!individual.isAnalyzed) {
```

Comment:
I think in addition to consuming honey, this function should also consume a reasonable amount of energy, to make it similar to other "scanning" components in the mod such as the geolyzer.